Initial Extending the Prisma Client documentation

matthewmueller opened the pull request:
## Describe this PR

- <s>[Extending the Prisma Client](https://deploy-preview-2602--prisma2-docs.netlify.app/docs/guides/database/extend-client)</s>
- [Custom Validation](https://deploy-preview-2602--prisma2-docs.netlify.app/docs/concepts/components/prisma-client/custom-validation)
- [Computed Fields](https://deploy-preview-2602--prisma2-docs.netlify.app/docs/concepts/components/prisma-client/computed-fields)
- [Excluding Fields](https://deploy-preview-2602--prisma2-docs.netlify.app/docs/concepts/components/prisma-client/excluding-fields)
- [Custom Models](https://deploy-preview-2602--prisma2-docs.netlify.app/docs/concepts/components/prisma-client/custom-models)

<!-- Please describe the issue, problem or reason for the PR... A little context helps us scan the PR first, without having to dive into the code -->
This PR explains how to extend the Prisma Client to support validation, computed fields, excluding fields and organize / add custom logic. This documentation describes how to add these features yourself, but eventually we'd like to bake these features into the Prisma Client itself. This will hopefully fill a gap until we know how to and are able to.

## Changes

- Added 1 guide and 4 new pages to concepts

<!-- What changes have you made? Keep it brief!

- Refactored example to include new database field
- ... -->

## What issue does this fix?

<!-- Link the issue this is solving, if applicable, using the issue number. This can be found to the right of the issue title, or in the url.

Fixes #1234  -->

- Addresses: https://github.com/prisma/prisma/issues/7161

matthewmueller commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
```diff
@@ -0,0 +1,400 @@
+---
+title: 'Validation, Computed Fields, and Custom Logic'
```
Not too sure about the organization. I feel like this fits more in Concepts, but unfortunately, I wrote this like a Guide.  Reviewing the website, I'm a bit worried that putting it in Guide's under a generic title like "How to Extend the Prisma Client" will make it undiscoverable.

I'm happy to slice and dice the content to make it fit more into Concepts. Otherwise, we call it a Guide with a better title like "How to add runtime validation, computed fields and custom logic to your app". This could even be a couple guides.

tanberry replied:
@matthewmueller and @janpio this is indeed a legit challenge!

The restructuring I am looking at (too slowly!) tries to answer this problem, with a version of what Jan suggests. The Concepts section is for high-level concepts, the WHY and the WHAT. Then under the Guides is the HOW content, the procedural steps. I haven't looked closely at this one yet, but so far it seems we could have most of this in a new guide (under Guides) and then under concepts, perhaps on [this page](https://www.prisma.io/docs/concepts/components/prisma-client), we link to the guide about these advanced steps of working with the Client.

tanberry replied:
@matthewmueller another thought... what do you think of separating this PR into two PRs... that way we can go ahead and merge the ones we know are good to go (like updating the Preview features lists), and have the new file (052-extend-client.mdx) in a separate PR where we can work on our questions about it?

matthewmueller replied:
> (like updating the Preview features lists)

Oh shoot. This is a git mistake. This happened to me recently before. I pulled in other people's commits. I'll clean that up. This PR should only be about the Extending Prisma Client Guide. 

I plan to add 1 (or more) other pages in Concepts like Validation, Computed Fields, Excluding Fields and Custom Logic that points to sections of the Guide. And then eventually we can replace each of those links with something that Prisma natively does.

Jolg42 commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
```diff
@@ -0,0 +1,401 @@
+
+To kick things off, let's discuss how to abstract Prisma Client queries into functions. This will be used in the examples that follow.
+
+Let's build a signup function that requires user and team information to create rows in your database. We'll build this function first in Javascript then in Typescript.
```
```suggestion
Let's build a signup function that requires user and team information to create rows in your database. We'll build this function first in JavaScript then in TypeScript.
```
TODO: fix everywhere

matthewmueller replied:
Thanks

Jolg42 commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
```diff
@@ -0,0 +1,401 @@
+
+### Helper Function in Javascript
+
+In Javascript, that would look something like following:
```
We could probably use a code switcher here

janpio replied:
Top level language switcher like in "Getting Started" or just tabs?

Jolg42 replied:
Tabs for index.js / index.ts would be the best I guess, and could be done in the rest of the doc, also removing the lines that say `In Javascript, that would look something like following:`

matthewmueller replied:
Good call. Cleans up a lot of the article 🙌

Jolg42 commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
```diff
@@ -0,0 +1,401 @@
+
+Some data must be validated at runtime. For example, there's no way to type-check that your user's email contains an `@` symbol at build time. To ensure that the data you get is the data you expect, you can use runtime validation.
+
+The Node.js ecosystem provides a number of high-quality, easy-to-use validation libraries you can choose from including: [joi](https://github.com/sideway/joi), [validator.js,](https://github.com/validatorjs/validator.js) [yup](https://github.com/jquense/yup) and [superstruct](https://github.com/ianstormtaylor/superstruct).
```
```suggestion
The Node.js ecosystem provides a number of high-quality, easy-to-use validation libraries you can choose from including: [joi](https://github.com/sideway/joi), [zod](https://github.com/colinhacks/zod), [validator.js](https://github.com/validatorjs/validator.js), [yup](https://github.com/jquense/yup) and [superstruct](https://github.com/ianstormtaylor/superstruct).
```
We should add zod in here I think

Jolg42 commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
````diff
@@ -0,0 +1,401 @@
+The most obvious way to wrap the Prisma Client is by wrapping it in a `Model` class:
+
+```tsx highlight=4,23,28-30;normal
+import { PrismaClient } from '@prisma/client'
````
```suggestion
import { PrismaClient } from '@prisma/client'
import { assert } from 'superstruct'
```
missing import for assert

Jolg42 commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
````diff
@@ -0,0 +1,401 @@
+But what if you don't want to hide existing functionality? In this case, you can use `Object.assign` to extend the Prisma Client without limiting it's functionality:
+
+```tsx highlight=2,27-29;normal
+function Model(model: PrismaClient['user']) {
````
```suggestion
import { PrismaClient } from '@prisma/client'
import { assert } from 'superstruct'

function Model(model: PrismaClient['user']) {
```
missing imports?

matthewmueller replied:
Thanks, it gets a bit long to include everything, but I've added some ellipsis to explain the missing pieces.

tobiasdiez commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
```diff
@@ -0,0 +1,401 @@
+   */
+  async signup(input: Signup) {
+    assert(input, Signup)
+    const team = await prisma.team.create({
```
I guess this should use `model` instead of `prisma` directly, or otherwise there is no point in passing the prisma client as a constructor argument, or is there?

tobiasdiez commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
````diff
@@ -0,0 +1,401 @@
+```tsx highlight=4,23,28-30;normal
+import { PrismaClient } from '@prisma/client'
+
+class Model {
````
Maybe rename "Model" to "UserService"? I would expect a user model to be a custom representation of a user (with name, email etc as properties) but not a wrapper around prisma client.

matthewmueller replied:
Terminology is very tricky here. From Prisma's perspective a model represents a table or collection you can query on and what you get back is a model result. 

I like your definition of a model a bit more, which matches my understanding that a model is an item in a collection that has properties and methods. 

I find UserService overly generic, but since I don't have any better ideas, I'm going to cheat a bit and pluralize the function to avoid definition altogether.

```ts
async function main() {
  const prisma = new PrismaClient()
  const users = Users(prisma.user)
  const user = await users.signup(/* { ... } */)
  const numUsers = await users.count()
  console.log(user, numUsers)
}
```

Thanks for this comment :-)

tobiasdiez reviewed (changes requested):
Thanks for working on this.
In general it was an interesting read, and summarizes really well how one currently can work-around some of the limitations of prisma.

What I'm missing in the "Computed Fields" section is the strategy to wrap the "User" object in a class that re-exposes email, firstName, lastName but not password; and provides a method (or property) to get the fullName. This is what I would call a custom `UserModel`. This would "collect" the different `compute*` and `exclude` in one class.

janpio commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
```diff
@@ -0,0 +1,401 @@
+
+## Conclusion
+
+Today we've shown how you can implement some patterns on top of the Prisma Client to extend functionality and cover your data needs.
```
Maybe this could/should be a blog post instead of a docs page? Then a docs page that tells people that this is not possible yet could link to the blog (instead of the Guides section) to find the current workarounds.

matthewmueller replied:
Ended up writing a page for each. Some copy and paste, but tried to isolate the concepts. I don't have a huge preference on guide vs blog post, just kept it as a guide for now.

janpio reviewed (commented):
Content itself makes a lot of sense on a superficial read.
Left a few comments about structure etc.

camelmasa commented on content/300-guides/050-database/052-extend-client.mdx:
```diff
@@ -0,0 +1,445 @@
+
+Some data must be validated at runtime. For example, there's no way to type-check that your user's email contains an `@` symbol at build time. To ensure that the data you get is the data you expect, you can use runtime validation.
+
+The Node.js ecosystem provides a number of high-quality, easy-to-use validation libraries you can choose from including: [joi](https://github.com/sideway/joi), [validator.js,](https://github.com/validatorjs/validator.js) [yup](https://github.com/jquense/yup), [zod](https://github.com/colinhacks/zod) and [superstruct](https://github.com/ianstormtaylor/superstruct).
```
@matthewmueller 

If you already decided prisma client's embedded validator (joi, yup, zod, superstruct or etc), Could you write that here ?

matthewmueller replied:
Prisma currently has no opinions about custom validation, so you can choose any of the options above.

camelmasa replied:
Got it. Thanks for replying.

Jolg42 commented on content/200-concepts/100-components/01-prisma-schema/06-relations/420-referential-integrity.mdx:
```diff
@@ -35,37 +60,63 @@ generator js {
+| Cascade            | no         | yes        |
+| NoAction           | no         | yes        |
+
+Currently Prisma only implements the referential actions. Foreign keys also create constraints, which make it impossible to manipulate data in a way that would violate these constraints: instead of executing the query, the database responds with an error. These constraints will not be created if you emulate referential integrity in the client, so if you set the referential action to [`NoAction`]( ...
```
```suggestion
Currently Prisma only implements some referential actions (see table). Foreign keys also create constraints, which make it impossible to manipulate data in a way that would violate these constraints: instead of executing the query, the database responds with an error. These constraints will not be created if you emulate referential integrity in the client, so if you set the referential action to [`NoAction`](/concepts/components/prisma-schema/relations/referential-actions#noaction) there will be no checks to prevent you from breaking the referential integrity.
```
Maybe?

Jolg42 commented on content/200-concepts/100-components/01-prisma-schema/06-relations/425-referential-actions/100-special-rules-for-referential-actions.mdx:
```diff
@@ -1,15 +1,17 @@
-If the referential actions on the foreign key are set to something other than `NO ACTION`, the server will check for cycles or multiple cascade paths and return an error when executing the SQL.
+- Microsoft SQL Server doesn't allow cascading [referential actions](./) on a foreign key, if the relation chain causes a cycle or multiple cascade paths. If the referential actions on the foreign key are set to something other than `NO ACTION` (or `NoAction` if Prisma is managing referential integrity), the server will check for cycles or multiple cascade paths and return an error when executing ...
+
+- with MongoDB, using referential actions in Prisma requires that for any data model with self-referential relations or cycles between three models, you must set the referential action of `NoAction` to prevent the referential action emulations from looping infinitely. Be aware that by default, the `referentialIntegrity = "prisma"` mode is used for MongoDB, which means that Prisma manages [referen ...
```
```suggestion
- with MongoDB, using referential actions in Prisma requires that for any data model with self-referential relations or cycles between three models, you must set the referential action of `NoAction` to prevent the referential action emulations from looping infinitely. Be aware that by default, the `referentialIntegrity = "prisma"` mode is used for MongoDB, which means that Prisma manages [referential integrity](/concepts/components/prisma-schema/relations/referential-integrity).
```

tanberry replied:
@Jolg42 your edit does sound better is some ways (if spoken for example), but in tech docs we do wants to use the simple present tense form of "manages." @matthewmueller your choice, as author, but I'd vote for leaving it as you had it. :-)

Jolg42 commented on content/200-concepts/100-components/01-prisma-schema/06-relations/425-referential-actions/index.mdx:
```diff
@@ -353,7 +348,10 @@ If you are using either the [`delete()`](../../../../../guides/database/advanced
 Make sure you are happy with every case of `onDelete: Cascade` in your schema. If not, either:
 
-- Modify your Prisma schema and `db push` or `dev migrate` to change the database*or*
+- Modify your Prisma schema and `db push` or `dev migrate` to change the database
```
```suggestion
- Modify your Prisma schema and `db push` or `migrate dev` to change the database
```
typo

janpio reviewed (commented):
Uh, something went wrong with the merges I fear - the PR diff on GH now shows a _lot_ of unrelated changes from other commits and PRs that are not related to this actual change (and Joel for example made suggestions to some of these). Probably want to rebase or open a new PR for just your real changes.

matthewmueller commented on content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx:
```diff
@@ -0,0 +1,401 @@
+---
+title: 'Validation, Computed Fields, and Custom Logic'
```
Fixed

matthewmueller commented:
Okay, this is ready for one more review:

- Moved "Extending the Prisma Client" to guides
- Added 4 new independent pages to the docs: validation, computed fields, excluding fields, custom models. I tried to change up each sentence a bit, but some of it's definitely copy and paste (will we get penalized by google?)

nikolasburk commented on content/300-guides/050-database/052-extend-client.mdx:
```diff
@@ -0,0 +1,445 @@
+
+This example also demonstrates how you can use the built-in `Omit` type to exclude fields from the input, further restraining what the signup function will accept.
+
+<TabbedContent tabs={[<FileWithIcon text="TypeScript" icon="code"/>, <FileWithIcon text="JavaScript (Node.js)" icon="code"/>]}>
```
I'd suggest to drop the "Node.js" here:

```suggestion
<TabbedContent tabs={[<FileWithIcon text="TypeScript" icon="code"/>, <FileWithIcon text="JavaScript" icon="code"/>]}>
```

nikolasburk reviewed (commented):
This is really awesome content, great work Matt!!

I noted a few more stylistic things that I think @tanberry and @keerlu should consider before merging:

- We usually omit the article before "Prisma Client" (it's considered a proper noun), so it would be e.g. "Extending Prisma Client" instead of "Extending the Prisma Client"
- Some of the line highlights didn't seem to be correct, e.g. the first code snippet in the section "Extending the Prisma Client" where a highlight is placed on closing braces and end empty line
- In general, I found the use of line highlights not necessarily very helpful (just my personal opinion) since I didn't feel that the highlighted lines were always particularly more important than the rest of the snippet and it hasn't been pointed out in the text why something was highlighted
- I believe we use regular sentence casing instead of title casing for sections in the docs, so it would be e.g. "Excluding sensitive fields" instead of "Excluding Sensitive Fields"

thebiglabasky commented on content/200-concepts/100-components/02-prisma-client/063-excluding-fields.mdx:
```diff
@@ -0,0 +1,69 @@
+  ...keys: Key[]
+): Omit<User, Key> {
+  for (let key of keys) {
+    delete user[key]
```
Nit: Doesn't that risk of warning if the key doesn't exist in `user`?

matthewmueller replied:
The `Key extends keyof User` bit removes the warning, telling Typescript that Key is a key within User. It's pretty magical.

tanberry commented on content/200-concepts/100-components/02-prisma-client/064-custom-models.mdx:
````diff
@@ -0,0 +1,93 @@
+```
+
+Here we're only exposing a `signup` method from the Prisma Client. With this approach, you'll no longer be able to call Prisma Client methods like `findMany` or `upsert`. This approach works well when you have a large application and you want to intentionally limit what your models can do.
+
````
By 'Here" do you mean the section above, right? 
Should we briefly explain why wrapping `user` model in the `User` class disallows you from call the methods mentioned? Does wrapping mean that only what the class allows is available, and the only thing that class allows/provides is the signup feature? Or maybe this is super-obvious to even newbie developers?

matthewmueller replied:
Just tried to clarify a bit

tanberry commented on content/200-concepts/100-components/02-prisma-client/064-custom-models.mdx:
```diff
@@ -0,0 +1,93 @@
+## Extending the Prisma Client
+
+But what if you don't want to hide existing functionality? In this case, you can use `Object.assign` to extend the Prisma Client without limiting it's functionality:
+
```
Consider  something like "But what if you don't want to hide existing functionality, but still need to group some logic components?" 

(The reader might have already forgotten that the goal is to group logic functions/components....)

matthewmueller replied:
Thanks yep!

tanberry commented on content/200-concepts/100-components/02-prisma-client/064-custom-models.mdx:
```diff
@@ -0,0 +1,93 @@
+
+## Going Further
+
+- This is an excerpt from a larger guide on [How to Extend the Prisma Client](/guides/database/extend-client). In the guide we talk about how to add custom validation and computed fields to your model.
```
I am confused about the info architecture.... I see you link to the larger guide... you could also link to the individual topics about validation and computed fields, right?

tanberry commented on content/300-guides/050-database/052-extend-client.mdx:
```diff
@@ -0,0 +1,445 @@
+Let's say we also want to derive the user's full name from their first and last name. In the example below, you'll create a helper function to compute a full name from the first and last name stored in the database.
+
+### Computing a Full Name
+
```
How is this comment diff from what is in computed-fields.mdx? (My info architecture confusion...) Or are you doing some sort of fancy {include} and single-sourcing content?

tanberry reviewed (commented):
This is such valuable info! Let's talk about your plan for info architecture, but the content looks great.

matthewmueller commented on content/200-concepts/100-components/02-prisma-client/062-computed-fields.mdx:
```diff
@@ -0,0 +1,80 @@
+## Deriving a Full Name from a First and Last Name
+
+The Prisma Client does not yet natively support computed fields, but with a bit of TypeScript magic, you can define a function that accepts a generic as an input then extend that generic to ensure it conforms to a specific structure. Finally you can return that generic with additional computed fields. Let's see how that might look:
+
```
Yep, generic can be a noun :)

matthewmueller commented on content/300-guides/050-database/052-extend-client.mdx:
```diff
@@ -0,0 +1,445 @@
+The Node.js ecosystem provides a number of high-quality, easy-to-use validation libraries you can choose from including: [joi](https://github.com/sideway/joi), [validator.js,](https://github.com/validatorjs/validator.js) [yup](https://github.com/jquense/yup), [zod](https://github.com/colinhacks/zod) and [superstruct](https://github.com/ianstormtaylor/superstruct).
+
+Let's go with superstruct for the example below and extend the signup helper function to ensure that the data you get meets your expectation.
+
```
Thanks, also applies to Zod and Yup

matthewmueller commented on content/200-concepts/100-components/02-prisma-client/064-custom-models.mdx:
```diff
@@ -0,0 +1,93 @@
+## Wrap a Prisma Model in a Class
+
+In the example below, we're wrapping the `user` model in the Prisma Client within a `Users` class. This class adds a custom signup function.
+
```
Thanks, I tried making it a bit more explicit.

matthewmueller commented:
Merging. Thanks everyone!
